File: heattouls/autostart.py
"""Windowsログオン時の自動起動登録(HKCUのRunキー)。"""

import sys
import winreg
from pathlib import Path

RUN_KEY = r"Software\Microsoft\Windows\CurrentVersion\Run"
VALUE_NAME = "HeatTouls"
LEGACY_VALUE_NAME = "toulstudio"   # 旧名で登録済みの分を掃除するため


def _command() -> str:
    """コンソールを出さずに起動するコマンドライン。"""
    if getattr(sys, "frozen", False):
        # exeとしてビルド済み。自分自身を起動すればよい。
        return f'"{Path(sys.executable).resolve()}" --minimized'
    pythonw = Path(sys.executable).with_name("pythonw.exe")
    exe = pythonw if pythonw.exists() else Path(sys.executable)
    entry = Path(__file__).resolve().parent.parent / "main.py"
    return f'"{exe}" "{entry}" --minimized'


def status():
    """登録済みならコマンド文字列、未登録ならNone。"""
    try:
        with winreg.OpenKey(winreg.HKEY_CURRENT_USER, RUN_KEY) as key:
            value, _type = winreg.QueryValueEx(key, VALUE_NAME)
            return value
    except FileNotFoundError:
        return None


def enable() -> str:
    command = _command()
    with winreg.CreateKey(winreg.HKEY_CURRENT_USER, RUN_KEY) as key:
        winreg.SetValueEx(key, VALUE_NAME, 0, winreg.REG_SZ, command)
        # 旧名で登録されたままだと二重に起動してしまうので消しておく
        try:
            winreg.DeleteValue(key, LEGACY_VALUE_NAME)
        except FileNotFoundError:
            pass
    return command


def disable() -> bool:
    removed = False
    try:
        with winreg.OpenKey(winreg.HKEY_CURRENT_USER, RUN_KEY, 0, winreg.KEY_SET_VALUE) as key:
            for name in (VALUE_NAME, LEGACY_VALUE_NAME):
                try:
                    winreg.DeleteValue(key, name)
                    removed = True
                except FileNotFoundError:
                    pass
    except FileNotFoundError:
        return False
    return removed

File: src/HeatTouls/Core/Autostart.cs
using Microsoft.Win32;

namespace HeatTouls.Core;

/// <summary>Windowsログオン時の自動起動登録(HKCUのRunキー)。</summary>
public static class Autostart
{
    private const string RunKey = @"Software\Microsoft\Windows\CurrentVersion\Run";
    private const string ValueName = "HeatTouls";
    private const string LegacyValueName = "toulstudio";   // 旧名で登録済みの分を掃除するため

    /// <summary>コンソールを出さずに起動するコマンドライン。</summary>
    private static string BuildCommand()
    {
        // 発行された exe をそのまま指す。WinExe なのでコンソールは開かない。
        var exe = Environment.ProcessPath ?? Environment.GetCommandLineArgs()[0];
        return $"\"{Path.GetFullPath(exe)}\" --minimized";
    }

    /// <summary>登録済みならコマンド文字列、未登録なら null。</summary>
    public static string? Status()
    {
        using var key = Registry.CurrentUser.OpenSubKey(RunKey);
        return key?.GetValue(ValueName) as string;
    }

    public static string Enable()
    {
        var command = BuildCommand();
        using var key = Registry.CurrentUser.CreateSubKey(RunKey, writable: true);
        key.SetValue(ValueName, command, RegistryValueKind.String);
        // 旧名で登録されたままだと二重に起動してしまうので消しておく
        key.DeleteValue(LegacyValueName, throwOnMissingValue: false);
        return command;
    }

    public static bool Disable()
    {
        using var key = Registry.CurrentUser.OpenSubKey(RunKey, writable: true);
        if (key is null)
        {
            return false;
        }

        var removed = false;
        foreach (var name in new[] { ValueName, LegacyValueName })
        {
            if (key.GetValue(name) is not null)
            {
                key.DeleteValue(name, throwOnMissingValue: false);
                removed = true;
            }
        }
        return removed;
    }
}
